File: back-end/app/models/ingredient_model.py
from pydantic import BaseModel, Field
from typing import Optional

class NutritionInfo(BaseModel):
    calories: float = Field(..., description="Calories in kcal", json_schema_extra={"example": 33.81})
    fat: float = Field(..., description="Total fat in grams", json_schema_extra={"example": 0.08})
    carbs: float = Field(..., description="Carbohydrates in grams", json_schema_extra={"example": 8.87})
    protein: float = Field(..., description="Protein in grams", json_schema_extra={"example": 0.35})

class Ingredient(BaseModel):
    food_name: str = Field(..., json_schema_extra={"example": "grape"})
    serving_qty: float = Field(..., json_schema_extra={"example": 10})
    serving_unit: str = Field(..., json_schema_extra={"example": "grapes"})
    serving_weight_grams: float = Field(..., json_schema_extra={"example": 49})
    nutrition: NutritionInfo
    food_group: Optional[int] = Field(
        default=None,
        description="Food category (1=Dairy, 2=Protein, 3=Fruit, 4=Vegetable, 5=Meat, 6=Grain, 7=Fat, 8=Other)",
        json_schema_extra={"example": 3}
    )
    image: Optional[str] = Field(
        default=None,
        json_schema_extra={"example": "https://nix-tag-images.s3.amazonaws.com/586_thumb.jpg"}
    )

File: front-end/src/models/ingredient.model.ts
export class NutritionInfo {
    calories: number;
    fat: number;
    carbs: number;
    protein: number;

    constructor(data: {
        calories: number,
        fat: number,
        carbs: number,
        protein: number
    }) {
        this.calories = data.calories;
        this.fat = data.fat;
        this.carbs = data.carbs;
        this.protein = data.protein;
    }
}

export class Ingredient {
    food_name: string;
    serving_qty: number;
    serving_unit: string;
    serving_weight_grams: number;
    nutrition: NutritionInfo;
    image?: string;

    constructor(data: {
        food_name: string,
        serving_qty: number,
        serving_unit: string,
        serving_weight_grams: number,
        nutrition: NutritionInfo,
        image?: string
    }) {
        this.food_name = data.food_name;
        this.serving_qty = data.serving_qty;
        this.serving_unit = data.serving_unit;
        this.serving_weight_grams = data.serving_weight_grams;
        this.nutrition = new NutritionInfo(data.nutrition);
        this.image = data.image;
    }

    // Helper method to create from API response
    static fromJson(json: any): Ingredient {
        return new Ingredient({
            food_name: json.food_name,
            serving_qty: json.serving_qty,
            serving_unit: json.serving_unit,
            serving_weight_grams: json.serving_weight_grams,
            nutrition: {
                calories: json.nutrition.calories,
                fat: json.nutrition.fat,
                carbs: json.nutrition.carbs,
                protein: json.nutrition.protein
            },
            image: json.image
        });
    }
}